src/toil/cwl/cwltoil.py

```diff
@@ -515,17 +513,22 @@ def toil_make_tool(toolpath_object, loading_context):
 
 
 class ToilFsAccess(cwltool.stdfsaccess.StdFsAccess):
-    """Custom filesystem access class which handles toil filestore references.
-
-    """
-
-    def __init__(self, basedir, file_store=None):
+    """Custom filesystem access class which handles toil filestore references."""
+    def __init__(self, basedir: str, file_store: AbstractFileStore = None):
         self.file_store = file_store
         super(ToilFsAccess, self).__init__(basedir)
 
     def _abs(self, path):
+        """Used to fetch a path to determine if a file exists so this should not error on missing files."""
+        # Link to where this is called in cwltool:
+        # https://github.com/common-workflow-language/cwltool/blob/beab66d649dd3ee82a013322a5e830875e8556ba/cwltool/stdfsaccess.py#L43
         if path.startswith("toilfs:"):
-            return self.file_store.readGlobalFile(FileID.unpack(path[7:]))
+            try:
+                return self.file_store.readGlobalFile(FileID.unpack(path[7:]))
+            except NoSuchFileException:
+                # normally toil initiates a similar filename and then places the file there
+                # here we return the created filename even if the file was not there
+                return self.file_store.getLocalTempFileName()
```

Comment:
At this point we only have a `toilfs:` uri (which the class inheriting this doesn't understand), we want a temp uri path returned that's guaranteed to not exist if it's also not present in the filestore.

Unfortunately this method is called as `os.path.exists(_abs(path))`, so the `_abs(path)` portion cannot error on a missing file before hitting the `os.path.exists()`.  It's also called on pretty much every other file operation: https://github.com/common-workflow-language/cwltool/blob/beab66d649dd3ee82a013322a5e830875e8556ba/cwltool/stdfsaccess.py#L40

Perhaps another way of implementing this is to also overwrite `StdFsAccess.exists()`?  I was attempting to reproduce the behavior of cwltool however, since `_abs()` never raises there and is called for a bunch of other functions as well.

Reply:
So when the file isn't supposed to exist all we need to do is return a path to a nonexistent file?

Reply:
I've adjusted it.